### ISSUE TITLE
Add support for external OpenGL textures

### ISSUE THREAD

**Issue by Drakirus** [go-flutter-desktop/go-flutter#102]
The flutter engine ABI supports external OpenGL textures allowing the embedder to draw to OpenGL directly. [engine#7087](https://github.com/flutter/engine/pull/7087)

It's needed to implement the [VideoPlayer](https://github.com/flutter/plugins/tree/master/packages/video_player).

**Comment by gedw99** [comment]
Will also be useful for doing 3D graphics and webrtc.

**Comment by cloudwebrtc** [comment]
@Drakirus @gedw99 
I have done a texture patch on flutter/engine and it seems to work very well, but I don't know how to port it to go-flutter. Here is my Texture rendering code, maybe it will help you.
https://github.com/cloudwebrtc/flutter-engine/blob/texture_for_glfw/shell/platform/glfw/external_texture_gl.cc

![image](https://user-images.githubusercontent.com/17733746/56503593-79fa2c80-6548-11e9-91b7-83293f8b932b.png)

**Comment by Drakirus** [comment]
@cloudwebrtc Is the `EventChannel` require for the Texture to work?

**Comment by cloudwebrtc** [comment]
@Drakirus
I think EventChannel is needed, because when the video frame changes in size and aspect ratio, the flutter side needs to adjust the display scale or fit, EventChannel is used to send notifications to the flutter side.
In addition, if we want to adapt WebRTC.PeerConnection, or other plugins with many event sources, EventChannel is needed.

**Comment by Drakirus** [comment]
Hi @cloudwebrtc I'm seeing progress on flutter/engine about the textureAPI :+1:.  
I'm trying to implement the textureAPI for go-flutter based your CPP code. 

While implementing the API, I hit a wall.  

My understanding was that you register a texture using: `FlutterEngineRegisterExternalTexture`, then mark a new frame with `FlutterDesktopMarkExternalTextureFrameAvailable`. At some point, in between or after those calls, the flutter-engine calls you back using the `gl_external_texture_frame_callback`. Am my missing something?
Here is my progress on the implementation: [commit](https://github.com/go-flutter-desktop/go-flutter/commit/e494cb017a94a65480f8a7bcdabcc135da3cadbb)

**Comment by cloudwebrtc** [comment]
@Drakirus  Sorry to reply so late, I see you in the code RegisterExternalTexture(textureId: 400), and after a while MarkExternalTextureFrameAvailable(textureId: 600), this place textureId should be a fixed value of int64_t when registering once, for example: textureId = 0, when your The plugin generates a new video frame, only need to execute MarkExternalTextureFrameAvailable(texture: 0), then the flutter engine will call GLExternalTextureFrameCallback(texture: 0, ...).

**Comment by Drakirus** [comment]
@cloudwebrtc Thanks for your reply!
The testing code that I pushed was wrong, I've updated it: [commit](https://github.com/go-flutter-desktop/go-flutter/compare/feature/gl_texture).

I'm still facing the same issue: `gl_external_texture_frame_callback` never gets called. :disappointed: 

I've updated the code to use a bare C handler [code](https://github.com/go-flutter-desktop/go-flutter/compare/feature/gl_texture#diff-1d5e1cd32a9c5dc395fb2d7a38b91108R21), `OnAcquireExternalTexture` never get called.

The terminal output looks great:
```
[....]
Texture: register result: true
go-flutter: no handler found for channel flutter/isolate
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
Texture: New Frame result: true
go-flutter: closing application
```

**Comment by cloudwebrtc** [comment]
@Drakirus Your changes are correct, just add Texture(textureId: 1) to flutter lib/main.dart.
I tested it under linux:
```
[ERROR:flutter/shell/platform/embedder/embedder.cc(578)] Could not create external texture.
=== OnAcquireExternalTexture: 1
Unknown pixel config 0x0
[ERROR:flutter/shell/platform/embedder/embedder.cc(578)] Could not create external texture.
=== OnAcquireExternalTexture: 1
Unknown pixel config 0x0
[ERROR:flutter/shell/platform/embedder/embedder.cc(578)] Could not create external texture.
=== OnAcquireExternalTexture: 1
Unknown pixel config 0x0
[ERROR:flutter/shell/platform/embedder/embedder.cc(578)] Could not create external texture.
=== OnAcquireExternalTexture: 1
Unknown pixel config 0x0
[ERROR:flutter/shell/platform/embedder/embedder.cc(578)] Could not create external texture.
=== OnAcquireExternalTexture: 1
Unknown pixel config 0x0
[ERROR:flutter/shell/platform/embedder/embedder.cc(578)] Could not create external texture.
Texture: New Frame result: true
=== OnAcquireExternalTexture: 1
Unknown pixel config 0x0
[ERROR:flutter/shell/platform/embedder/embedder.cc(578)] Could not create external texture.
go-flutter: closing application

```

**Comment by Drakirus** [comment]
`Texture(textureId: 1)` that what I was missing, Thanks!!

**Comment by Drakirus** [comment]
Making progress: 
![2019-07-24_00-30](https://user-images.githubusercontent.com/7476655/61751924-424e8b00-adaa-11e9-8294-d8acc2560193.png)

A demo is available at: https://github.com/go-flutter-desktop/examples/tree/feature/gl_texture/texture_tutorial, use `hover run -b '@feature/gl_texture'` to try it out.

@cloudwebrtc everything was implemented thanks to you, big thanks!

**Comment by cloudwebrtc** [comment]
Wow, that's awesome.

**Comment by Drakirus** [comment]
![2019-07-25_17-11](https://user-images.githubusercontent.com/7476655/61886106-5ce64880-aeff-11e9-8ecb-8b676c6af0d2.png)
Updated the demo. Added an very simple videoplayer

**Comment by Drakirus** [comment]
fix taged in tag: [v0.29.0](https://github.com/go-flutter-desktop/go-flutter/releases/tag/v0.29.0)
